src/server/scripts/Northrend/Naxxramas/boss_four_horsemen.cpp

```diff
@@ -82,315 +99,510 @@ enum FourHorsemen
-                            Thane->Respawn();
+                    cBoss->DespawnOrUnsummon(0u);
+                    cBoss->SetRespawnTime(15u);
+                }
+                else
+                {
+                    TC_LOG_WARN("scripts", "FourHorsemenAI: Encounter resetting but horseman with id %d is not present", uint32(boss));
+                }
+        }
 
-                        if (!Lady->IsAlive())
-                            Lady->Respawn();
+        void EncounterCleared()
+        {
+            if (instance->GetBossState(BOSS_HORSEMEN) == DONE)
+                return;
+            instance->SetBossState(BOSS_HORSEMEN, DONE);
+            //instance->DoUpdateAchievementCriteria(ACHIEVEMENT_CRITERIA_TYPE_BE_SPELL_TARGET, SPELL_ENCOUNTER_CREDIT);
+            DoCastAOE(SPELL_ENCOUNTER_CREDIT, true);
+        }
 
-                        if (!Baron->IsAlive())
-                            Baron->Respawn();
+        void EnterCombat(Unit* who) override
+        {
+            if (instance->GetBossState(BOSS_HORSEMEN) != NOT_STARTED) // another horseman already did it
+                return;
+            BeginEncounter();
+        }
 
-                        if (!Sir->IsAlive())
-                            Sir->Respawn();
+        void EnterEvadeMode() override
+        {
+            ResetEncounter();
+        }
 
-                        ENSURE_AI(boss_four_horsemen::boss_four_horsemenAI, Thane->AI())->encounterActionReset = true;
-                        ENSURE_AI(boss_four_horsemen::boss_four_horsemenAI, Lady->AI())->encounterActionReset = true;
-                        ENSURE_AI(boss_four_horsemen::boss_four_horsemenAI, Baron->AI())->encounterActionReset = true;
-                        ENSURE_AI(boss_four_horsemen::boss_four_horsemenAI, Sir->AI())->encounterActionReset = true;
+        void Reset() override
+        {
+            if (!me->IsAlive())
+                return;
+            _myMovementFinished = false;
+            _nextMovement = 0;
+            _timeDied = 0;
+            _ourMovementFinished = false;
+            me->SetReactState(REACT_AGGRESSIVE);
+            SetCombatMovement(false);
+            me->SetCombatPulseDelay(0);
+            me->ResetLootMode();
+            events.Reset();
+            summons.DespawnAll();
+        }
 
-                        ENSURE_AI(boss_four_horsemen::boss_four_horsemenAI, Thane->AI())->EnterEvadeMode();
-                        ENSURE_AI(boss_four_horsemen::boss_four_horsemenAI, Lady->AI())->EnterEvadeMode();
-                        ENSURE_AI(boss_four_horsemen::boss_four_horsemenAI, Baron->AI())->EnterEvadeMode();
-                        ENSURE_AI(boss_four_horsemen::boss_four_horsemenAI, Sir->AI())->EnterEvadeMode();
-                    }
+        void JustDied(Unit* /*killer*/) override
+        {
+            if (instance->GetBossState(BOSS_HORSEMEN) != IN_PROGRESS) // necessary in case a horseman gets one-shot
+                return BeginEncounter();
```

Comment:
please use "return;" when the return type is void otherwise it might look like the called function is returning some kind of value.